File: python/cad.py
"""Shared headless noBS CAD session helpers for the tools in this directory."""
import base64
import json
import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from mcp_client import McpClient  # noqa: E402

SERVER = os.environ.get("NBCAD_MCP", "nbcad-mcp")  # the headless noBS CAD MCP executable; a bare name is looked up on PATH


def session(log_path):
    return McpClient(SERVER, log_path=str(log_path))


def scene_summary(client):
    scene = client.call("solid_scene")
    bodies = scene.get("bodies", [])
    out = {"bodies": len(bodies), "errors": scene.get("errors", []), "body_details": []}
    for body in bodies:
        faces = body.get("faces", [])
        planar = sum(1 for f in faces if f.get("plane"))
        mesh = body.get("mesh") or {}
        verts = mesh.get("positions") or mesh.get("vertices")
        bbox = None
        if verts:
            flat = verts if isinstance(verts[0], (int, float)) else [c for v in verts for c in v]
            xs, ys, zs = flat[0::3], flat[1::3], flat[2::3]
            bbox = [round(max(xs) - min(xs), 3), round(max(ys) - min(ys), 3), round(max(zs) - min(zs), 3)]
        out["body_details"].append({"id": body.get("id"), "faces": len(faces), "planar_faces": planar, "cylindrical_faces": sum(1 for f in faces if f.get("cylinder")), "bbox": bbox})
    return out


def holes_from_scene(client):
    """Vertical holes as (x, y from the body's bounding-box minimum, diameter, counterbore)."""
    scene = client.call("solid_scene")
    compare = client.call("cad_compare_solids")
    body = (compare.get("bodies") or [{}])[0]
    mn = body.get("bbox_min", [0, 0, 0])
    groups = []
    for b in scene.get("bodies", []):
        for f in b.get("faces", []):
            c = f.get("cylinder")
            if not c:
                continue
            axis, origin, r = c.get("axis") or {}, c.get("origin") or {}, c.get("radius")
            if abs(axis.get("z", 0)) < 0.9 or r is None:
                continue
            x, y = origin["x"], origin["y"]
            for g in groups:
                if abs(g["x"] - x) < 0.3 and abs(g["y"] - y) < 0.3:
                    g["radii"].append(r)
                    break
            else:
                groups.append({"x": x, "y": y, "radii": [r]})
    holes = []
    for g in sorted(groups, key=lambda g: (g["y"], g["x"])):
        rad = sorted(g["radii"])
        holes.append({"x": round(g["x"] - mn[0], 2), "y": round(g["y"] - mn[1], 2), "diameter": round(2 * rad[0], 2), "counterbore": round(2 * rad[-1], 2) if len(rad) > 1 else None})
    return {"bbox_min": mn, "bbox_max": body.get("bbox_max"), "holes": holes}


def export_step(client, path):
    data = base64.b64decode(client.call("solid_export_step", {})["bytes_base64"])
    Path(path).write_bytes(data)
    return {"path": str(path), "bytes": len(data), "is_step": data[:9] == b"ISO-10303"}

File: lib/cad.js
/**
 * Headless noBS CAD helpers shared by the tools: run a script, export STEP,
 * re-import a STEP, and summarise the scene (bounding box, faces, vertical holes).
 */
import { readFileSync, writeFileSync } from 'node:fs'
import { basename } from 'node:path'
import { McpClient } from './mcp-client.js'

function stripExtension(path) {
  return path.replace(/\.[^./\\]+$/, '')
}

/** Run `fn` against a fresh engine session whose stderr goes to `logPath`. */
export async function withSession(server, logPath, fn) {
  const client = new McpClient(server, { logPath })
  try {
    await client.ready
    return await fn(client)
  } finally {
    await client.close()
  }
}

export async function sceneSummary(client) {
  const scene = await client.call('solid_scene')
  const bodies = scene.bodies ?? []
  const out = { bodies: bodies.length, errors: scene.errors ?? [], body_details: [] }
  for (const body of bodies) {
    const faces = body.faces ?? []
    const mesh = body.mesh ?? {}
    const verts = mesh.positions ?? mesh.vertices
    let bbox = null
    if (Array.isArray(verts) && verts.length) {
      const flat = typeof verts[0] === 'number' ? verts : verts.flat()
      const axis = (k) => { let lo = Infinity, hi = -Infinity; for (let i = k; i < flat.length; i += 3) { if (flat[i] < lo) lo = flat[i]; if (flat[i] > hi) hi = flat[i] } return round(hi - lo, 3) }
      bbox = [axis(0), axis(1), axis(2)]
    }
    out.body_details.push({ id: body.id, faces: faces.length, planar_faces: faces.filter((f) => f.plane).length, cylindrical_faces: faces.filter((f) => f.cylinder).length, bbox })
  }
  return out
}

function round(value, digits) {
  const k = 10 ** digits
  return Math.round(value * k) / k
}

/** Vertical holes as (x, y from the body's bounding-box minimum, diameter, counterbore). */
export async function holesFromScene(client) {
  const scene = await client.call('solid_scene')
  const compare = await client.call('cad_compare_solids')
  const body = (compare.bodies ?? [])[0] ?? {}
  const mn = body.bbox_min ?? [0, 0, 0]
  const groups = []
  for (const b of scene.bodies ?? []) {
    for (const f of b.faces ?? []) {
      const c = f.cylinder
      if (!c) continue
      const axis = c.axis ?? {}
      const origin = c.origin ?? {}
      const r = c.radius
      if (Math.abs(axis.z ?? 0) < 0.9 || r === undefined || r === null) continue
      const x = origin.x, y = origin.y
      const group = groups.find((g) => Math.abs(g.x - x) < 0.3 && Math.abs(g.y - y) < 0.3)
      if (group) group.radii.push(r)
      else groups.push({ x, y, radii: [r] })
    }
  }
  groups.sort((a, b) => a.y - b.y || a.x - b.x)
  const holes = groups.map((g) => {
    const rad = [...g.radii].sort((a, b) => a - b)
    return { x: round(g.x - mn[0], 2), y: round(g.y - mn[1], 2), diameter: round(2 * rad[0], 2), counterbore: rad.length > 1 ? round(2 * rad[rad.length - 1], 2) : null }
  })
  return { bbox_min: mn, bbox_max: body.bbox_max ?? null, holes }
}

export async function exportStep(client, path) {
  const result = await client.call('solid_export_step', {})
  const data = Buffer.from(result.bytes_base64, 'base64')
  writeFileSync(path, data)
  return { path, bytes: data.length, is_step: data.subarray(0, 9).toString('latin1') === 'ISO-10303' }
}

function message(error) {
  return String(error?.message ?? error).slice(0, 4000)
}

/** Run a version 1 .nbcad.jsonc script in a blank document and export STEP; the same JSON the model has always seen. */
export async function runScript(server, scriptPath, stepPath) {
  const result = { ok: false, script: scriptPath, step: stepPath }
  try {
    const source = readFileSync(scriptPath, 'utf8')
    await withSession(server, stripExtension(stepPath) + '.mcp.log', async (client) => {
      const run = await client.call('cad_interface', { action: 'script', source, mode: 'fast' })
      Object.assign(result, { ok: true, steps_completed: run.steps_completed, checks_completed: run.checks_completed, elapsed_ms: run.elapsed_ms })
      result.scene = await sceneSummary(client)
      result.holes = await holesFromScene(client)
      result.export = await exportStep(client, stepPath)
    })
  } catch (error) {
    result.ok = false
    result.error = message(error)
  }
  return result
}

/** Re-import a STEP file and report the same summary. */
export async function inspectStep(server, stepPath) {
  const result = { ok: false, step: stepPath }
  try {
    await withSession(server, stripExtension(stepPath) + '.inspect.log', async (client) => {
      const catalog = await client.interface('catalog')
      const groups = {}
      for (const g of catalog.groups ?? []) for (const op of g.operations ?? []) groups[op] = g.id
      await client.interface('execute', { group: groups.solid_import_step, operation: 'solid_import_step', arguments: { file_name: basename(stepPath), data_base64: readFileSync(stepPath).toString('base64') } })
      result.ok = true
      result.scene = await sceneSummary(client)
      result.holes = await holesFromScene(client)
    })
  } catch (error) {
    result.ok = false
    result.error = message(error)
  }
  return result
}
